Public/Src/Cache/MemoizationStore/Interfaces/Sessions/Selector.cs

```diff
@@ -24,6 +26,18 @@ public Selector(ContentHash contentHash, byte[] output = null)
             Output = output;
         }
 
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="Selector" /> struct.
+        /// </summary>
+        public Selector(BinaryReader reader)
+        {
+            Contract.Requires(reader != null);
+            ContentHash = new ContentHash(reader);
+
+            var length = reader.ReadInt32();
+            Output = length == -1 ? null : reader.ReadBytes(length);

```

Comment:
We should extract method to something: SerializeNullableArray and DeserializeNullableArray.

I've noticed the pattern like this a lot.